torchtext/datasets/text_classification.py

```diff
@@ -0,0 +1,416 @@
+
+    def __getitem__(self, i):
+        return self.examples[i]
+
+    def __len__(self):
+        try:
+            return len(self.examples)
+        except TypeError:
+            return 2**32
+
+    def __iter__(self):
+        for x in self.examples:
+            yield x
+
+
+class TextClassificationDataset(TextDataset):
+    """Defines text classification datasets.
+        Currently, we only support the following datasets:
+
+             - AG_NEWS
+             - SogouNews
+             - DBpedia
+             - YelpReviewPolarity
+             - YelpReviewFull
+             - YahooAnswers
+             - AmazonReviewPolarity
+             - AmazonReviewFull
+
+    """
+
+    def __init__(self, url, root='.data',
+                 text_field=None, label_field=None, ngrams=1):
+        """Create a text classification dataset instance.
+
+        Arguments:
+            dataset_name: The name of dataset, include "ag_news", "sogou_news",
+                "dbpedia", "yelp_review_polarity", "yelp_review_full",
+                "yahoo_answers", "amazon_review_full", "amazon_review_polarity".
+            root: Directory where the dataset are saved. Default: ".data"
+            text_field: The field that will be used for the sentence. If not given,
+                'spacy' token will be used.
+            label_field: The field that will be used for the label. If not given,
+                'float' token will be used.
+            ngrams: a contiguous sequence of n items from s string text.
+                Default: 1
+
+        Examples:
+            >>> url = 'https://drive.google.com/uc?export=download&id=0Bz8a_Dbh9QhbUDNpeUdjb0wxRms'
+            >>> txt_cls = TextClassificationDataset(url, ngrams=2)
+
+        """
+        fields = []
+        fields.append(('text', text_field if text_field is not None
+                       else data.Field(tokenize=data.get_tokenizer('spacy'),
+                                       init_token='<SOS>',
+                                       eos_token='<EOS>')))
+        fields.append(('label', label_field if label_field is not None
+                       else data.LabelField(dtype=torch.float)))
+
+        super(TextClassificationDataset, self).__init__([], dict(fields))
+
+        self.dataset_name = self.__class__.__name__
+        self.root = root
+        self.raw_folder = os.path.join(self.root, self.__class__.__name__, 'raw')
+        self.processed_folder = os.path.join(self.root,
+                                             self.__class__.__name__,
+                                             'processed')
+        filepath = os.path.join(self.processed_folder, self.dataset_name + '.train')
+        if not os.path.isfile(filepath):
+            download(self.url, self.raw_folder, self.dataset_name)
+            preprocess(self.raw_folder, self.processed_folder, self.dataset_name)
+        self.train_examples = load_text_classification_data(filepath, self.fields, ngrams)
```

Comment:
Since it's all being read into memory I'd expect the user to just pass in this already from memory. So this doesn't need to take a path, but can take two lists. We can then add support for paths later on.

Reply:
Fixed.